controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,418 @@
+		r.discoveryLastRefreshed.Add(time.Minute*10).Before(time.Now()) {
+		err := r.refreshDiscoveryInfo()
+		r.discoveryLastRefreshed = time.Now()
+
+		if err != nil {
+			return "", err
+		}
+	}
+
+	for _, resc := range r.apiResourceList {
+		groupVerison, err := schema.ParseGroupVersion(resc.GroupVersion)
+		if err != nil {
+			r.Log.Error(err, "Cannot Parse Group and version in getApiVersion ", "GroupVersion:", resc.GroupVersion)
+
+			continue
+		}
+
+		group := groupVerison.Group
+		version := groupVerison.Version
+		// Consider groupversion == v1 or groupversion == app1/v1
+		for _, apiResource := range resc.APIResources {
+			if apiResource.Kind == kind && group == apiGroup {
+				return version, nil
+			}
+		}
+	}
+
+	if !skipRefresh {
+		// Get new discoveryInfo, when any resource is not found
+		err := r.refreshDiscoveryInfo()
+		if err != nil {
+			return "", err
+		}
+
+		r.discoveryLastRefreshed = time.Now()
+		// Retry one more time after refresh the discovery
+		return r.getAPIVersion(kind, apiGroup, true)
+	}
+
+	return "", ErrNotFoundDiscovery
+}
+
+// Retrieve all groups and versions to add in config sync
+// Constraints present only kind and group so this function helps to find the version
+func (r *ConstraintStatusReconciler) refreshDiscoveryInfo() error {
+	discoveryClient := r.ClientSet.Discovery()
+
+	apiList, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		return err
+	}
+
+	r.apiResourceList = apiList
+
+	return nil
+}
+
+func (r *ConstraintStatusReconciler) getUniqSyncOnly() []v1alpha1.SyncOnlyEntry {
+	syncOnlySet := map[v1alpha1.SyncOnlyEntry]bool{}
+	// Add to table for unique filtering
+	r.constraintMap.Range(func(ckey interface{}, entries interface{}) bool {
+		constraintName, ok := ckey.(string)
+		if !ok {
+			r.Log.V(1).Info("Error to parse constraint name")
+
+			return true
+		}
+
+		arrayEntries, ok := entries.([]v1alpha1.SyncOnlyEntry)
+		if !ok {
+			r.Log.V(1).Info("Error to parse Constraint SyncOnlyEntry values", "constraintName:", constraintName)
+
+			return true
+		}
+
+		for _, entry := range arrayEntries {
+			syncOnlySet[entry] = true
+		}
+
+		return true
+	})
+
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+	for key := range syncOnlySet {
+		syncOnlys = append(syncOnlys, key)
+	}
+
+	return syncOnlys
+}
+
+func (r *ConstraintStatusReconciler) updateConfigSyncOnlyEntry(
+	ctx context.Context, config *v1alpha1.Config, syncOnlys []v1alpha1.SyncOnlyEntry,
+) error {
+	config.Spec.Sync.SyncOnly = syncOnlys
+
+	return r.Update(ctx, config, &client.UpdateOptions{})
+}
+
+func (r *ConstraintStatusReconciler) addConstraintToMap(ckey string, entries []v1alpha1.SyncOnlyEntry) {
+	r.constraintMap.Store(ckey, entries)
+}
+
+// This function works when constraintPodStatus deleted event.
+// Collect all ConstraintTemplates and refresh r.constraintMap with current constraint
+func (r *ConstraintStatusReconciler) handleDeleteEvent(ctx context.Context, config *v1alpha1.Config) error {
```

Comment:
The controller runs concurrent reconciles and this deletes all constraint entries. I think you could run into race conditions where another reconcile is adding something and this is wiping it out.

Reply:
I have no idea except this way Do you have an idea to replace this

Reply:
key = contraintPodstatus name then everything is beautiful